tox.ini

```diff
@@ -1,144 +1,91 @@
 recreate = True
 
 [testenv]
-commands = {envpython} -m pytest {posargs:tests/unittests}
+basepython = python3
 setenv =
     LC_ALL = en_US.utf-8
 passenv=
     PYTEST_ADDOPTS
 
-[testenv:flake8]
-basepython = python3
+[flake_env]
+envdir = {toxworkdir}/.flake_env
 deps =
     flake8==3.9.2
-commands = {envpython} -m flake8 {posargs:cloudinit/ tests/ tools/ setup.py}
-
-# https://github.com/gabrielfalcao/HTTPretty/issues/223
-setenv =
-    LC_ALL = en_US.utf-8
-
-[testenv:pylint]
-basepython = python3
-deps =
-    # requirements
     pylint==2.11.1
-    # test-requirements because unit tests are now present in cloudinit tree
+    black==21.12b0
+    isort==5.10.1
     -r{toxinidir}/test-requirements.txt
     -r{toxinidir}/integration-requirements.txt
+
+[testenv:flake8]
+envdir = {[flake_env]envdir}
+deps = {[flake_env]deps}
+commands = {envpython} -m flake8 {posargs:cloudinit/ tests/ tools/ setup.py}
+
+[testenv:pylint]
+envdir = {[flake_env]envdir}
+deps = {[flake_env]deps}
 commands = {envpython} -m pylint {posargs:cloudinit tests tools}
 
 [testenv:black]
-basepython = python3
-deps =
-    # requirements
-    black==21.12b0
+envdir = {[flake_env]envdir}
+deps = {[flake_env]deps}
 commands = {envpython} -m black . --check
 
 [testenv:isort]
-basepython = python3
-deps =
-    isort==5.10.1
+envdir = {[flake_env]envdir}
+deps = {[flake_env]deps}
 commands = {envpython} -m isort . --check-only
 
 [testenv:format]
-basepython = python3
-deps =
-    black==21.12b0
-    isort==5.10.1
+envdir = {[flake_env]envdir}
+deps = {[flake_env]deps}
 commands =
     {envpython} -m isort .
     {envpython} -m black .
 
 [testenv:py3]
-basepython = python3
 deps =
     -r{toxinidir}/test-requirements.txt
 commands = {envpython} -m pytest \
             --durations 10 \
             {posargs:--cov=cloudinit --cov-branch \
             tests/unittests}
 
-[testenv:py27]
-basepython = python2.7
-deps = -r{toxinidir}/test-requirements.txt
+[testenv:lowest-supported]
+# Tox is going to install requirements from pip. This is fine for
+# testing python version compatibility, but when we build cloud-init, we are
+# building against the dependencies in the OS repo, not pip. The OS
+# dependencies will generally be older than what is found in pip.
 
-[flake8]
-# E203: whitespace before ':', doesn't adhere to pep8 or black formatting
-# W503: line break before binary operator
-ignore=E203,W503
-exclude = .venv,.tox,dist,doc,*egg,.git,build,tools
-per-file-ignores =
-    cloudinit/cmd/main.py:E402
+# To obtain these versions, check the versions of these libraries
+# in the oldest support Ubuntu distro.
+deps =
+    jinja2==2.10
+    oauthlib==2.0.6
+    pyserial==3.4
+    configobj==5.0.6
+    pyyaml==3.12
+    requests==2.18.4
+    jsonpatch==1.16
+    jsonschema==2.6.0
+    netifaces==0.10.4
+    # test-requirements
+    httpretty==0.8.14
+    pytest==3.3.2
+    pytest-cov==2.5.1
+    attrs==17.4.0  # Needed by pytest and default causes failures
```

Comment:
Inline comments still break on bionic tox. This comment needs to be a separate line otherwise we'll hit something like:
```bash
ERROR: Invalid requirement: 'attrs==17.4.0 #  Needed by pytest and default causes failures'
```